[masonry] Implement Subgrid tests to verify functionality

stwrt opened the pull request:
#### f334b00ec9585dead30d247c23c946fe14ebf226
<pre>
[masonry] Implement Subgrid tests to verify functionality
<a href="https://bugs.webkit.org/show_bug.cgi?id=303513">https://bugs.webkit.org/show_bug.cgi?id=303513</a>
<a href="https://rdar.apple.com/problem/165806666">rdar://problem/165806666</a>

Reviewed by Elika Etemad.

While going through the subgrid tests it was difficult to make out exactly what the original was hoping to test exactly.
The tests are from the original proposal and have just been tweaked over time to follow the new changes to the spec.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/grid-lanes/tentative/subgrid/waterfall/line-numbers-001-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/grid-lanes/tentative/subgrid/waterfall/line-numbers-001-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/grid-lanes/tentative/subgrid/waterfall/line-numbers-001.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/grid-lanes/tentative/subgrid/waterfall/line-numbers-002-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/grid-lanes/tentative/subgrid/waterfall/line-numbers-002-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/grid-lanes/tentative/subgrid/waterfall/line-numbers-002.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/grid-lanes/tentative/subgrid/waterfall/named-lines-001-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/grid-lanes/tentative/subgrid/waterfall/named-lines-001-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/grid-lanes/tentative/subgrid/waterfall/named-lines-001.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/grid-lanes/tentative/subgrid/waterfall/named-lines-002-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/grid-lanes/tentative/subgrid/waterfall/named-lines-002-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/grid-lanes/tentative/subgrid/waterfall/named-lines-002.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/grid-lanes/tentative/subgrid/waterfall/named-lines-003-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/grid-lanes/tentative/subgrid/waterfall/named-lines-003-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/grid-lanes/tentative/subgrid/waterfall/named-lines-003.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/grid-lanes/tentative/subgrid/waterfall/sizing-001-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/grid-lanes/tentative/subgrid/waterfall/sizing-001-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/grid-lanes/tentative/subgrid/waterfall/sizing-001.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/grid-lanes/tentative/subgrid/waterfall/sizing-002-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/grid-lanes/tentative/subgrid/waterfall/sizing-002-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/grid-lanes/tentative/subgrid/waterfall/sizing-002.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/grid-lanes/tentative/subgrid/waterfall/sizing-003-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/grid-lanes/tentative/subgrid/waterfall/sizing-003-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/grid-lanes/tentative/subgrid/waterfall/sizing-003.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/grid-lanes/tentative/subgrid/waterfall/w3c-import.log: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/grid-lanes/tentative/subgrid/waterfall/wrapping-001-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/grid-lanes/tentative/subgrid/waterfall/wrapping-001-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/grid-lanes/tentative/subgrid/waterfall/wrapping-001.html: Added.

Canonical link: <a href="https://commits.webkit.org/303975@main">https://commits.webkit.org/303975@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2bb09acad2a520684633618ee92693c591447183

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/134099 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/6610 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/45306 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/141679 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/86161 "Built successfully") | ⏳ 🛠 ios-apple 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/7144 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/6474 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/102562 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/69853 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/8c839b7d-41fa-4c69-b188-d1e46c18677e) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/137046 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/4993 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/120208 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/83357 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/1b320fa1-8ba1-46a0-816d-982f8f930d7b) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/4872 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/2506 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/1494 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/114043 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/38332 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/144325 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/6280 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/38911 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/110930 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/6362 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/5304 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/111161 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28203 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/4738 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/116470 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/60050 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/6332 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/34701 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/6178 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/6423 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/6286 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

fantasai commented on LayoutTests/imported/w3c/web-platform-tests/css/css-grid/grid-lanes/tentative/subgrid/waterfall/display-001.html:
```diff
@@ -0,0 +1,73 @@
+.grid {
+  display: inline-grid-lanes;
+  grid-lanes-direction: row;
+  grid-template-columns: 50px 100px 150px;
```
Small nit: Suggest factoring this out into a .explicit class and putting the grid-template-columns: auto auto auto declarations into a .auto class, and that'll make the two variants parallel and easier to see that they're alternates of each other.

fantasai replied:
Other thought: maybe rename this `display-` series to `sizing-` since that's really what it's about, the display types being variations on a sizing test fundamentally.

fantasai commented on LayoutTests/imported/w3c/web-platform-tests/css/css-grid/grid-lanes/tentative/subgrid/waterfall/line-numbers-001.html:
```diff
@@ -0,0 +1,92 @@
+  <div class="subgrid" style="grid-column: 2 / 5">
+    <div class="subitem" style="grid-column: 1">1</div>
+    <div class="subitem" style="grid-column: 2">2</div>
+    <div class="subitem" style="grid-column: 3">3</div>
```
Scramble the order of these so that the test doesn't accidentally pass if they are auto-placed. Ditto the rest of this series and the `named-lines` series.

fantasai commented on LayoutTests/imported/w3c/web-platform-tests/css/css-grid/grid-lanes/tentative/subgrid/waterfall/named-lines-001.html:
```diff
@@ -0,0 +1,81 @@
+<!-- Subgrid with named lines, items placed using those names -->
+<div class="grid">
+  <div class="item" style="grid-column: 1">A</div>
+  <div class="subgrid" style="grid-column: 2 / 5">
```
Make this auto-placed by giving it only a `span 3`. That will check that it works even when it's auto-placed. (Parent line names won't work for an auto-placed subgrid, but the subgrid's own names should work.)

fantasai commented on LayoutTests/imported/w3c/web-platform-tests/css/css-grid/grid-lanes/tentative/subgrid/waterfall/named-lines-002.html:
```diff
@@ -0,0 +1,93 @@
+<!-- Subgrid at b-e, items use parent's line names -->
+<div class="grid">
+  <div class="item" style="grid-column: a">A</div>
+  <div class="subgrid" style="grid-column: b / e">
```
Create a variant of this test (`named-lines-003`) where the subgrid is auto-placed, and test that the parent line names are ignored.

fantasai commented on LayoutTests/imported/w3c/web-platform-tests/css/css-grid/grid-lanes/tentative/subgrid/waterfall/named-lines-002.html:
```diff
@@ -0,0 +1,93 @@
+  <div class="subgrid" style="grid-column: b / e">
+    <div class="subitem" style="grid-column: b">b</div>
+    <div class="subitem" style="grid-column: c">c</div>
+    <div class="subitem" style="grid-column: d">d</div>
```
Have the subgrid add its own names, and use one of these items to check that it is additive (i.e. we can reference the subgrid's names _or_ the parent names, both work).

fantasai reviewed (commented):
Overall these look really really good! Left some comments on specifics.

I'm having trouble understanding what you're trying to do with the `placement-` series, let's talk about it.

For the record: future work to expand these tests (not suggesting to do this now) includes:
- For the `display-` series (really about sizing), add asymmetric margins/border/padding/gaps to everything.
- For the `line-numbers-`/`named-lines-` series, create variations for the different `display` nesting scenarios.
- Also for the `line-numbers-`/`named-lines-` series, add some out-of-range placement to check behavior in the error case.
- For the `named-lines-` series, create a variant that is auto-placed, since that won't inherit the line names.

fantasai commented on LayoutTests/imported/w3c/web-platform-tests/css/css-grid/grid-lanes/tentative/subgrid/waterfall/named-lines-003.html:
```diff
@@ -0,0 +1,83 @@
+  <div class="subgrid" style="grid-column: span 3">
+    <div class="subitem" style="grid-column: b">b</div>
+    <div class="subitem" style="grid-column: c">c</div>
+    <div class="subitem" style="grid-column: d">d</div>
```
I think this isn't going to be able to fail, right? Because it'll give the same rendering whether the parent line names are ignored or honored.

stwrt replied:
Right, this should succeed because it is not randomly ordered.

fantasai replied:
Isn't that bad though? The goal is for the test to fail if the line names aren't ignored, right?

stwrt replied:
Addressed comment.

stwrt committed:
[masonry] Implement Subgrid tests to verify functionality
https://bugs.webkit.org/show_bug.cgi?id=303513
rdar://problem/165806666

Reviewed by Elika Etemad.

While going through the subgrid tests it was difficult to make out exactly what the original was hoping to test exactly.
The tests are from the original proposal and have just been tweaked over time to follow the new changes to the spec.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/grid-lanes/tentative/subgrid/waterfall/line-numbers-001-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/grid-lanes/tentative/subgrid/waterfall/line-numbers-001-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/grid-lanes/tentative/subgrid/waterfall/line-numbers-001.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/grid-lanes/tentative/subgrid/waterfall/line-numbers-002-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/grid-lanes/tentative/subgrid/waterfall/line-numbers-002-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/grid-lanes/tentative/subgrid/waterfall/line-numbers-002.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/grid-lanes/tentative/subgrid/waterfall/named-lines-001-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/grid-lanes/tentative/subgrid/waterfall/named-lines-001-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/grid-lanes/tentative/subgrid/waterfall/named-lines-001.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/grid-lanes/tentative/subgrid/waterfall/named-lines-002-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/grid-lanes/tentative/subgrid/waterfall/named-lines-002-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/grid-lanes/tentative/subgrid/waterfall/named-lines-002.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/grid-lanes/tentative/subgrid/waterfall/named-lines-003-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/grid-lanes/tentative/subgrid/waterfall/named-lines-003-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/grid-lanes/tentative/subgrid/waterfall/named-lines-003.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/grid-lanes/tentative/subgrid/waterfall/sizing-001-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/grid-lanes/tentative/subgrid/waterfall/sizing-001-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/grid-lanes/tentative/subgrid/waterfall/sizing-001.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/grid-lanes/tentative/subgrid/waterfall/sizing-002-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/grid-lanes/tentative/subgrid/waterfall/sizing-002-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/grid-lanes/tentative/subgrid/waterfall/sizing-002.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/grid-lanes/tentative/subgrid/waterfall/sizing-003-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/grid-lanes/tentative/subgrid/waterfall/sizing-003-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/grid-lanes/tentative/subgrid/waterfall/sizing-003.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/grid-lanes/tentative/subgrid/waterfall/w3c-import.log: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/grid-lanes/tentative/subgrid/waterfall/wrapping-001-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/grid-lanes/tentative/subgrid/waterfall/wrapping-001-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/grid-lanes/tentative/subgrid/waterfall/wrapping-001.html: Added.

Canonical link: https://commits.webkit.org/303975@main
